sw/host/opentitanlib/src/transport/ti50emulator/gpio.rs

```diff
@@ -1,32 +1,342 @@
+        }
+        bail!("GPIO stream insufficient amount of data");
+    }
+}
+
+impl fmt::Display for GpioConfiguration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?} {} {}", self.pin_mode, self.pull_mode, self.value)
+    }
+}
+
+/// Multi value logic with drive strength representations.
+/// (Subset of IEEE 1164).
+enum Logic {
+    StrongZero,
+    StrongOne,
+    WeekZero,
+    WeekOne,
+    HiImpedance,
+}
+
+impl fmt::Display for Logic {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Logic::StrongZero => write!(f, "0"),
+            Logic::StrongOne => write!(f, "1"),
+            Logic::WeekZero => write!(f, "L"),
+            Logic::WeekOne => write!(f, "H"),
+            Logic::HiImpedance => write!(f, "Z"),
+        }
+    }
+}
+
+impl From<GpioConfiguration> for Logic {
+    /// Convert the GPIO state stored in `state` to multi-valued logic.
+    fn from(state: GpioConfiguration) -> Logic {
+        match (state.pin_mode, state.pull_mode, state.value) {
+            (Some(PinMode::PushPull), _, false) => Logic::StrongZero,
+            (Some(PinMode::PushPull), _, true) => Logic::StrongOne,
+            (Some(PinMode::OpenDrain), _, false) => Logic::StrongZero,
+            (Some(PinMode::OpenDrain), PullMode::PullUp, true) => Logic::StrongZero,
+            (Some(PinMode::OpenDrain), PullMode::PullDown, true) => Logic::StrongZero,
+            (Some(PinMode::OpenDrain), PullMode::None, true) => Logic::HiImpedance,
+            (Some(PinMode::Input), PullMode::PullUp, _) => Logic::WeekOne,
+            (Some(PinMode::Input), PullMode::PullDown, _) => Logic::WeekZero,
+            (Some(PinMode::Input), PullMode::None, _) => Logic::HiImpedance,
+            (None, _, _) => Logic::HiImpedance,
+        }
+    }
+}
+
+/// The function calculates the state of the "virtual wire" that connects the "host" and DUT GPIO port.
+/// If any inconsistencies are detected between the sides, it returns an error with the problem description.
+/// The state on the host side is passed by `host` argument and on DUT side by `dut`.
+/// Parameter `name` contain name of GPIO pin.
+fn resolve_state(name: &str, host: &Logic, dut: &Logic) -> Result<bool> {
+    match (host, dut) {
+        (Logic::StrongOne, Logic::StrongZero) => {
+            Err(
+                GpioError::PinValueConflict(name.to_string(), host.to_string(), dut.to_string())
+                    .into(),
+            )
+        }
+        (Logic::StrongOne, _) => Ok(true),
+        (Logic::StrongZero, Logic::StrongOne) => {
+            Err(
+                GpioError::PinValueConflict(name.to_string(), host.to_string(), dut.to_string())
+                    .into(),
+            )
+        }
+        (Logic::StrongZero, _) => Ok(false),
+        (
+            Logic::WeekZero | Logic::WeekOne | Logic::HiImpedance,
+            Logic::StrongOne | Logic::WeekOne,
+        ) => Ok(true),
```

Comment:
Weak pull of the simulated chip under test is assumed stronger than the weak pull of the simulated environment (host).  That is how we want it, in order to be able to properly exercise strapping logic.  Maybe it is worth documenting our choice, so that a future maintainer can consider that before making changes.